identity-service/src/typed-routes/solana/solanaRelayChecks.ts

```diff
@@ -0,0 +1,349 @@
+import {
+  NATIVE_MINT,
+  ASSOCIATED_TOKEN_PROGRAM_ID,
+  TOKEN_PROGRAM_ID,
+  decodeInstruction,
+  isCloseAccountInstruction,
+  isTransferCheckedInstruction
+} from '@solana/spl-token'
+import {
+  PublicKey,
+  TransactionInstruction,
+  Secp256k1Program
+} from '@solana/web3.js'
+import audiusLibsWrapper from '../../audiusLibsInstance'
+import models from '../../models'
+import { InvalidRelayInstructionError } from './InvalidRelayInstructionError'
+import {
+  decodeAssociatedTokenAccountInstruction,
+  isCreateAssociatedTokenAccountInstruction,
+  isCreateAssociatedTokenAccountIdempotentInstruction
+} from './programs/associatedToken'
+import {
+  decodeClaimableTokenInstruction,
+  isTransferClaimableTokenInstruction
+} from './programs/claimable-tokens'
+import { decodeRewardManagerInstruction } from './programs/reward-manager'
+import config from '../../config'
+
+const MEMO_PROGRAM_ID = 'Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo'
+const CLAIMABLE_TOKEN_PROGRAM_ID: string = config.get(
+  'solanaClaimableTokenProgramAddress'
+)
+const REWARDS_MANAGER_PROGRAM_ID: string = config.get(
+  'solanaRewardsManagerProgramId'
+)
+const JUPITER_AGGREGATOR_V6_PROGRAM_ID =
+  'JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4'
+
+const audioMintAddress: string = config.get('solanaMintAddress')
+const usdcMintAddress: string = config.get('solanaUSDCMintAddress')
+
+const REWARD_MANAGER: string = config.get('solanaRewardsManagerProgramPDA')
+
+const claimableTokenAuthorities = [audioMintAddress, usdcMintAddress].reduce(
+  (acc, mint) => ({
+    ...acc,
+    [mint]: PublicKey.findProgramAddressSync(
+      [new PublicKey(mint).toBytes().slice(0, 32)],
+      new PublicKey(CLAIMABLE_TOKEN_PROGRAM_ID)
+    )[0]
+  }),
+  {} as Record<string, PublicKey>
+)
+
+/**
+ * Only allow the createTokenAccount instruction of the Associated Token
+ * Account program, provided it has matching close instructions.
+ * Close instructions are on the Token Programand are not validated here.
+ */
+const assertAllowedAssociatedTokenAccountProgramInstruction = (
+  instructionIndex: number,
+  instruction: TransactionInstruction,
+  instructions: TransactionInstruction[]
+) => {
+  const decodedInstruction =
+    decodeAssociatedTokenAccountInstruction(instruction)
+  if (
+    isCreateAssociatedTokenAccountInstruction(decodedInstruction) ||
+    isCreateAssociatedTokenAccountIdempotentInstruction(decodedInstruction)
+  ) {
+    const allowedMints = [NATIVE_MINT.toBase58(), usdcMintAddress]
+    const mintAddress = decodedInstruction.keys.mint.pubkey.toBase58()
+    if (!allowedMints.includes(mintAddress)) {
+      throw new InvalidRelayInstructionError(
+        instructionIndex,
+        `Mint not allowed for Associated Token Account program: ${mintAddress}`
+      )
+    }
+
+    // Protect against feePayer drain by ensuring that there's always as
+    // many account close instructions as creates
```

Comment:
Do we need to validate that the accounts in the close instructions match the ones in the create instructions? Or would the whole transaction fail if you tried to close the same account twice and leave one of the others open?
Or does that not make any sense? Are we always operating on just a single account here?

Reply:
Each create/close instruction operates on an individual account, but if there are multiple this code runs multiple times, finding matches and ensuring the number of creates == closes (at least that's the intent!) so hopefully it's doing what you're asking